runtime/metricsview/search_query.go

```diff
@@ -0,0 +1,292 @@
+			planRaw string
+			resRaw  string
+			attrRaw string
+		)
+		err = rows.Scan(&planRaw, &resRaw, &attrRaw)
+		if err != nil {
+			return nil, err
+		}
+
+		var plan []druid.QueryPlan
+		err = json.Unmarshal([]byte(planRaw), &plan)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(plan) == 0 {
+			return nil, fmt.Errorf("failed to parse policy filter")
+		}
+		if plan[0].Query.Filter == nil {
+			// if we failed to parse a filter we return and run UNION query.
+			// this can happen when the row filter is complex
+			// TODO: iterate over this and integrate more parts like joins and subfilter in policy filter
+			return nil, errDruidNativeSearchUnimplemented
+		}
+		query = *plan[0].Query.Filter
+	}
+	dsn, err := druid.GetDSN(olap.(drivers.Handle).Config())
+	if err != nil {
+		return nil, err
+	}
+	if dsn == "" {
+		return nil, fmt.Errorf("druid connector config not found in instance")
+	}
+
+	// Build a native query
+	limit := 100
+	if a.Root.Limit != nil {
+		limit = int(*a.Root.Limit)
+	}
+	dims := make([]string, 0)
+	virtualCols := make([]druid.NativeVirtualColumns, 0)
+	for _, f := range a.Root.DimFields {
+		dim, err := a.lookupDimension(f.Name, true)
+		if err != nil {
+			return nil, err
+		}
+		// if the dimension is a expression we need a virtual column that can be scanned in SearchDimensions
+		if dim.Expression != "" {
+			virtualCols = append(virtualCols, druid.NativeVirtualColumns{
+				Type:       "expression",
+				Name:       fmt.Sprintf("%v_virtual_native", f.Name), // The name of the virtual column should not clash with actual column
+				Expression: dim.Expression,
+			})
+			dims = append(dims, fmt.Sprintf("%v_virtual_native", f.Name))
+		} else {
+			dims = append(dims, trimQuotes(f.Expr))
+		}
+	}
+	req := druid.NewNativeSearchQueryRequest(trimQuotes(*a.Root.FromTable), q.Search, dims, virtualCols, limit, q.TimeRange.Start, q.TimeRange.End, query) // TODO: timestamps may be nil!
+
+	// Execute the native query
+	var res druid.NativeSearchQueryResponse
+	nq := druid.NewNativeQuery(druidSQLDSN.ReplaceAllString(dsn, "/v2/"))
+	err = nq.Do(ctx, req, &res, req.Context.QueryID, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the response to a SearchResult
+	result := make([]SearchResult, 0)
+	for _, re := range res {
+		for _, r := range re.Result {
+			result = append(result, SearchResult{
+				Dimension: strings.TrimSuffix(r.Dimension, "_virtual_native"),
+				Value:     r.Value,
+			})
+		}
+	}
+	return result, nil
+}
+
+type searchSQLBuilder struct {
+	ast    *AST
+	search string
+	out    *strings.Builder
+	args   []any
+}
+
+func searchSQL(ast *AST, search string) (string, []any, error) {
+	b := &searchSQLBuilder{
+		ast:    ast,
+		search: search,
+		out:    &strings.Builder{},
+	}
+
+	err := b.writeSelect(ast.Root)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return b.out.String(), b.args, nil
+}
+
+func (b *searchSQLBuilder) writeSelect(n *SelectNode) error {
+	for i, f := range n.DimFields {
```

Comment:
Looking at this, I wonder if it could be simplified by creating a `Query` and calling `NewAST` and `.SQL()` separately for *each* dimension and combining the outputs with `UNION ALL`?

That might help avoid needing to access the AST internals here (and also remove some duplicated SQL output logic).

Reply:
Actually that's what I had thought initially but the current AST sql lacks two features:
1. It can't add an alias to a plain dimension select.
2. It can't select constants
These are required to generate following query:
```
SELECT 
  column1 AS Value,
  'dimension_name1' AS dimension
FROM model
GROUP BY Value
WHERE ....
UNION ALL 
SELECT 
  column2 AS Value,
  'dimension_name2' AS dimension
FROM model
WHERE ...
GROUP BY Value
...
```

Now we can circumnavigate this limitation by querying each dimension separately as well but that will lead to one query per dimension.

We can also add the missing functionality to the AST.
Please let me know what do you think.

Reply:
Got it, I can think of two ways to work around this:
1. Wrap the generated SQL for each dimension with `SELECT 'dim name 1' as dim, dim1 as val FROM (<generated SQL>)`
2. After calling `NewAST`, you can manually override the ASTs fields like this:
```go
ast.Root.DimFields = append(ast.Root.DimFields, FieldNode{
  Name: "dimension",
  Expr: ast.Root.DimFields[0].Name,
})
ast.Root.DimFields[0].Name = "value"
```